src/runtime/d3d12compute.cpp

```diff
@@ -0,0 +1,2765 @@
+#endif
+
+// A Printer that automatically reserves stack space for the printer buffer:
+// (the managed printers in 'printer.h' rely on malloc)
+template<uint64_t length = 1024, int type = BasicPrinter>
+class StackPrinter : public Printer<type, length> {
+public:
+    StackPrinter(void *ctx = NULL) : Printer<type, length>(ctx, buffer) { }
+    StackPrinter& operator () (void *ctx = NULL) {
+        this->user_context = ctx;
+        return *this;
+    }
+    uint64_t capacity() const { return length; }
+private:
+    char buffer [length];
+};
+
+// v trace and logging utilities for debugging v
+// ! definitely not super thread-safe stuff... !
+// in case there's no 'user_context' available in the scope of a function:
+static void *const user_context = NULL;
+//
+#if HALIDE_D3D12_TRACE
+    #ifndef DEBUG_RUNTIME
+    #define debug(x) Printer<BasicPrinter>(x)
+    #endif//DEBUG_RUNTIME
+    static volatile int indent_lock = 0;
+    static const char indent_pattern [] = "   ";
+    static char  indent [2048] = { };
+    static int   indent_end   = 0;
+    struct TraceLogScope {
+        TraceLogScope() {
+            while (__sync_lock_test_and_set(&indent_lock, 1)) {}
+            for (const char *p = indent_pattern; *p; ++p) {
+                indent[indent_end++] = *p;
+            }
+            __sync_lock_release(&indent_lock);
+        }
+        ~TraceLogScope() {
+            while (__sync_lock_test_and_set(&indent_lock, 1)) {}
+            for (const char *p = indent_pattern; *p; ++p) {
+                indent[--indent_end] = '\0';
+            }
+            __sync_lock_release(&indent_lock);
+        }
+    };
+    #define TRACEINDENT     ((const char*)indent)
+    #define TRACEPRINT(msg) debug(user_context) << TRACEINDENT << msg;
+    #define TRACELOG        TRACEPRINT("[@]" << __FUNCTION__ << "\n"); TraceLogScope trace_scope___;
+#else
+    #define TRACEINDENT ""
+    #define TRACEPRINT(msg)
+    #define TRACELOG
+#endif
+//
+#define ERRORLOG    error(user_context) << TRACEINDENT
+// ^ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ^
+
+static const char* d3d12_debug_dump();
+
+#define d3d12_halt(...)             \
+    ERRORLOG << __VA_ARGS__ << "\n" \
+             << d3d12_debug_dump()  \
+             << "!!! HALT !!!\n"
+
+void *d3d12_load_library(const char *name) {
+    TRACELOG;
+    void *lib = halide_load_library(name);
+    if (lib) {
+        TRACEPRINT("Loaded runtime library '" << name << "' at location " << lib << "\n");
+    } else {
+        d3d12_halt("Unable to load runtime library: " << name);
+    }
+    return lib;
+}
+
+void *d3d12_get_library_symbol(void *lib, const char *name) {
+    TRACELOG;
+    void* symbol = halide_get_library_symbol(lib, name);
+    if (symbol) {
+        TRACEPRINT("Symbol '" << name << "' found at " << symbol << "\n");
+    } else {
+        d3d12_halt("Symbol not found: " << name);
+    }
+    return symbol;
+}
+
+#define d3d12_debug_break (*((volatile int8_t*)NULL) = 0)
+
+#ifndef UNUSED
```

Comment:
This is used precisely one place in the code and that usage could be replaced easily by "input_buffer->sizeInBytes" instead of assigning to a local.

Reply:
I like to map (re-assign) variables from Halide to D3D12 variable names, as it helps when I am browsing through D3D12 documentation and the code I am writing (or any other API, for the matter):
```
UINT SizeInBytes = input_buffer->sizeInBytes;
```
I had to use the `UNUSED` macro quite a few times while developing the back-end, so there's a chance it will be handy again in the future.
I will add a comment at `UNUSED(SizeInBytes);` explaining that it is there because `TRACEPRINT` might boil down to "nothing" and `SizeInBytes` is just for debugging/logging.